tests/rw_lock_test.c

```diff
@@ -0,0 +1,104 @@
+    struct aws_rw_lock rw_lock;
+    aws_rw_lock_init(&rw_lock);
+
+    ASSERT_SUCCESS(aws_rw_lock_wrlock(&rw_lock), "rw_lock acquire should have returned success.");
+    ASSERT_SUCCESS(aws_rw_lock_wrunlock(&rw_lock), "rw_lock release should have returned success.");
+
+    ASSERT_SUCCESS(aws_rw_lock_rdlock(&rw_lock), "rw_lock acquire should have returned success.");
+    ASSERT_SUCCESS(aws_rw_lock_rdunlock(&rw_lock), "rw_lock release should have returned success.");
+
+    aws_rw_lock_clean_up(&rw_lock);
+
+    return 0;
+}
+AWS_TEST_CASE(rw_lock_aquire_release_test, test_rw_lock_acquire_release)
+
+struct thread_rw_lock_data {
+    struct aws_rw_lock rw_lock;
+    volatile int counter;
+    int max_counts;
+    volatile int thread_fn_increments;
+};
+
+static void rw_lock_thread_fn(void *rw_lock_data) {
+    struct thread_rw_lock_data *p_rw_lock = (struct thread_rw_lock_data *)rw_lock_data;
+    int finished = 0;
+    while (!finished) {
+
+        aws_rw_lock_rdlock(&p_rw_lock->rw_lock);
+
+        if (p_rw_lock->counter != p_rw_lock->max_counts) {
+            int counter = p_rw_lock->counter + 1;
+
+            aws_rw_lock_rdunlock(&p_rw_lock->rw_lock);
+            aws_rw_lock_wrlock(&p_rw_lock->rw_lock);
+
+            p_rw_lock->counter = counter;
+            p_rw_lock->thread_fn_increments += 1;
+
+            aws_rw_lock_wrunlock(&p_rw_lock->rw_lock);
+            aws_rw_lock_rdlock(&p_rw_lock->rw_lock);
+
+            finished = p_rw_lock->counter == p_rw_lock->max_counts;
+        }
+        else {
+            finished = 1;
+        }
+
+        aws_rw_lock_rdunlock(&p_rw_lock->rw_lock);
+    }
+}
+
+static int test_rw_lock_is_actually_rw_lock(struct aws_allocator *allocator, void *ctx) {
+
+    struct thread_rw_lock_data rw_lock_data = {
+        .counter = 0,
+        .max_counts = 1000000,
+        .thread_fn_increments = 0,
+    };
+
+    aws_rw_lock_init(&rw_lock_data.rw_lock);
+
+    struct aws_thread thread;
+    aws_thread_init(&thread, allocator);
+    ASSERT_SUCCESS(aws_thread_launch(&thread, rw_lock_thread_fn, &rw_lock_data, 0), "thread creation failed with error %d", aws_last_error());
+    int finished = 0;
+    while (!finished) {
+
+        aws_rw_lock_rdlock(&rw_lock_data.rw_lock);
+
+        finished = rw_lock_data.counter == rw_lock_data.max_counts;
+
+        aws_rw_lock_rdunlock(&rw_lock_data.rw_lock);
+    }
+
+    ASSERT_SUCCESS(aws_thread_join(&thread), "Thread join failed with error code %d.", aws_last_error());
+    ASSERT_INT_EQUALS(rw_lock_data.thread_fn_increments, rw_lock_data.max_counts, "Thread 2 should have written all data");
+    ASSERT_INT_EQUALS(rw_lock_data.max_counts, rw_lock_data.counter, "Both threads should have written exactly the max counts.");
+
+    aws_thread_clean_up(&thread);
+    aws_rw_lock_clean_up(&rw_lock_data.rw_lock);
+
+    return 0;
+}
+AWS_TEST_CASE(rw_lock_is_actually_rw_lock_test, test_rw_lock_is_actually_rw_lock)
```

Comment:
I’d add in a test to hammer the rw lock with lots of readers and a couple writers.